rfc/rfc-0003-developer-utils.md

```diff
@@ -0,0 +1,74 @@
+# RFC-0003: A pod with utilities for better developer experience
+
+**Authors:** Swapnil Bawaskar
+
+**Status:**
+
+**Pull Request URL:** [#1362](https://github.com/projectriff/riff/pull/1362)
+
+**Superseded by:** N/A
+
+**Supersedes:** [RFC 0001](https://github.com/projectriff/riff/pull/1359)
+
+**Related:**
+
+
+## Problem
+A developer would like to invoke the function for which they just created the `core deployer` or the `knative deployer`.
+When the streaming runtime is installed, they would like to test the processor by getting events into the stream and look into the contents of the stream. There is no one step mechanism for accomplishing any of these currently.
+An [earlier proposal](https://github.com/projectriff/riff/pull/1359) was to provide some of this functionality in the riff cli, this proposal expands on [Mark's comment](https://github.com/projectriff/riff/pull/1359#discussion_r348617981).
+
+### Anti-Goals
+We will only address this problem for development/demos, not production, so topics like auth/authz are out of scope for this document.
+
+## Solution
+We will ask the developers to run a `riff-developer-utils` pod in their development cluster. This pod will bundle a few binaries/scripts that users can invoke using `kubectl exec`. Since the commands will be run in-cluster, users won't need to run `kubectl port-forward`.
+We will have the following commands to start with:
+1. **invoke-core:** To invoke the given core deployer.  
+The command takes the form:  
+    ```
+    invoke-core <deployer-name> -n <namespace> -c "<curl-params>"
+    ```
+    where `<curl-params>` are passed along to the post-request.
+1. **invoke-knative:** To invoke the given knative deployer.
+The command takes the form:  
+    ```
+    invoke-knative <deployer-name> -n <namespace> -c "<curl-params>"
+    ```
+    where `<curl-params>` are passed along to the post-request.
+
+1. **publish-stream:** To publish an event to the given stream.
+The command takes the form:
+    ```
+    publish-stream <stream-name> -n <namespace> --payload <payload-as-string> --content-type <content-type> --header "<header-name>: <header-value>"
+    ```
+    where `stream-name`, `payload` and `content-type` are mandatory and `header` can be used multiple times.
+1. **subscribe-stream:** To subscribe for events from the given stream.
+The command takes the form:
+    ```
+    subscribe-stream <stream-name> --offset <long-offset> --payload-as-string
+    ```
+    This will display all the events in the stream from the given offset in the following json format:
+    ```
+    {"payload": "base64 encoded user payload","content-type": "the content type of the message","headers": {"user provided header": "while publishing"}}
+    ```
+    The payload will be base64 encoded unless the `--payload-as-string` flag is present, in which case it will be displayed as a string.
```

Comment:
the flag name `--payload-as-string` isn't sitting well with me. The purpose of base64 encoding the payload is so that it can be emitted as a string. By not base64 encoding, the payload value may or may not be a string. Nobody wants binary data in their shell.

I suggest dropping the `--payload-as-string` flag and force consumers to decode the payload if they are interested in the value.

Reply:
Dropped, the payload will always be base64 encoded since that's the only safe output.